### DISCUSSION TITLE
How to evaluate a constant expression?

### DISCUSSION THREAD

**Discussion by leafypout** [swc-project/swc#7617] [Q&A]
I have a JavaScript expression like:

`(-2473 * -2 + 17 * 447 + -12264)`

I want to evaluate it to a value (f64) from an `Expr`. I found the `Evaluator` struct but I'm not sure how to use it. Here's what I've tried:

```rust
let mut evaluator = Evaluator::new(
                Module {
                    span: Default::default(),
                    body: vec![ModuleItem::Stmt(Stmt::Expr(ExprStmt {
                        span: Default::default(),
                        expr: Box::new(Expr::Bin(expr.clone())),
                    }))],
                    shebang: None,
                },

                Marks::new()
            );
let result = evaluator.eval(&Expr::Bin(expr));
```

However the result is `None`. How can I evaluate an expression like this? I want to do it in a visitor, so I don't have access to a `Module`, and the input is also a script, not a module.

**Comment by kdy1** [comment]
You need to clone the whole `Module`

**Reply by leafypout** [reply]
What about Marks, and do I need to do anything special if I'm using `swc::Compiler`?

I did try clone the `Module` in `visit_mut_program` but still the same result unfortunately
